File: mobile-app/lib/data/ble/fake_ble_transport.dart
// ignore_for_file: prefer_initializing_formals

import 'dart:async';

import '../../domain/entities/models.dart';
import '../../domain/validators/config_validator.dart';
import '../protocol/ble_uuids.dart';
import '../protocol/protocol_codecs.dart';
import 'ble_transport.dart';

enum FakeBleScenario {
  normal,
  serviceMissing,
  mtuTooSmall,
  protocolMajor2,
  pairingClosed,
  writeTimeout,
  busyThenOk,
  rangeError,
  storageError,
  disconnectOnWrite,
}

enum FakeRideProfile { idle, moving, paused, lowBattery }

class FakeBleTransport implements BleTransport {
  FakeBleTransport({
    this.scenario = FakeBleScenario.normal,
    FakeRideProfile profile = FakeRideProfile.moving,
  }) : _profile = profile;

  FakeBleScenario scenario;
  FakeRideProfile _profile;
  bool _connected = false;
  bool _sensorTest = false;
  int _busyResponses = 0;
  Timer? _telemetryTimer;
  StreamController<BleLinkState>? _linkController;
  final Map<String, StreamController<List<int>>> _notifications = {};

  DeviceConfig _config = DeviceConfig.defaults;
  Telemetry _telemetry = const Telemetry(
    structVersion: 1,
    flags: 35,
    speedX100: 2550,
    avgSpeedX100: 2200,
    maxSpeedX100: 3500,
    tripDistanceCm: 125000,
    movingTimeS: 1800,
    odometerM: 123456,
    batteryMv: 3900,
    batteryPct: 75,
    rideState: RideState.moving,
    revolutions: 500,
    lastPulseAgeMs: 250,
    seq: 42,
    sensorState: SensorState.ok,
    powerState: PowerState.active,
  );

  DeviceInfo get _deviceInfo => DeviceInfo(
    structVersion: 1,
    protoMajor: scenario == FakeBleScenario.protocolMajor2 ? 2 : 1,
    protoMinor: 0,
    hwRevision: 1,
    model: 'BIKECOMP-XIAO',
    fwVersion: '1.0.0-fake',
    serial: const [1, 2, 3, 4, 5, 6, 7, 8],
    uptimeS: 3600,
    resetReason: ResetReason.powerOn,
    bootCount: 42,
    flags: scenario == FakeBleScenario.pairingClosed ? 0x0F : 0x1F,
  );

  @override
  Stream<BleAdapterState> get adapterState =>
      Stream.value(BleAdapterState.ready);

  @override
  Stream<BleScanResult> scan() async* {
    await Future<void>.delayed(const Duration(milliseconds: 40));
    yield const BleScanResult(
      deviceId: 'FA:KE:BI:KE:00:01',
      name: 'BikeComp-FAKE',
      rssi: -58,
      bondState: BondState.bonded,
    );
  }

  @override
  Future<void> stopScan() async {}

  @override
  Stream<BleLinkState> connect(String deviceId) {
    final controller = StreamController<BleLinkState>.broadcast();
    _linkController = controller;
    Timer.run(() async {
      controller.add(BleLinkState.connecting);
      await Future<void>.delayed(const Duration(milliseconds: 40));
      _connected = true;
      _restartTelemetryTimer();
      controller.add(BleLinkState.connected);
    });
    return controller.stream;
  }

  @override
  Future<void> disconnect() async {
    _connected = false;
    final controller = _linkController;
    _linkController = null;
    if (controller != null && !controller.isClosed) {
      controller.add(BleLinkState.disconnected);
      await controller.close();
    }
  }

  void _requireConnected() {
    if (!_connected) throw StateError('Fake BLE device is disconnected');
  }

  @override
  Future<int> requestMtu(int mtu) async {
    _requireConnected();
    return scenario == FakeBleScenario.mtuTooSmall ? 23 : 247;
  }

  @override
  Future<BleDiscovery> discoverServices() async {
    _requireConnected();
    if (scenario == FakeBleScenario.serviceMissing) {
      return const BleDiscovery(<String>{});
    }
    return const BleDiscovery(BleUuids.requiredCharacteristics);
  }

  @override
  Future<List<int>> read(String characteristicUuid) async {
    _requireConnected();
    return switch (characteristicUuid) {
      BleUuids.deviceInfo => ProtocolCodecs.encodeDeviceInfo(_deviceInfo),
      BleUuids.telemetry => ProtocolCodecs.encodeTelemetry(
        _profiledTelemetry(),
      ),
      BleUuids.configRead => ProtocolCodecs.encodeConfig(_config),
      BleUuids.commandResult => ProtocolCodecs.encodeCommandResult(
        const CommandResult(
          structVersion: 1,
          commandId: DeviceCommandId.unknown,
          status: CommandStatus.ok,
          detail: 0,
          token: 0,
          payload: <int>[],
        ),
      ),
      _ => throw StateError(
        'Unsupported fake characteristic $characteristicUuid',
      ),
    };
  }

  @override
  Future<void> writeWithResponse(
    String characteristicUuid,
    List<int> value,
  ) async {
    _requireConnected();
    if (scenario == FakeBleScenario.disconnectOnWrite) {
      _connected = false;
      _linkController?.add(BleLinkState.disconnected);
      throw StateError('Injected disconnect');
    }
    if (characteristicUuid == BleUuids.configWrite) {
      await _writeConfig(value);
      return;
    }
    if (characteristicUuid == BleUuids.command) {
      await _writeCommand(value);
      return;
    }
    throw StateError('Unsupported fake write $characteristicUuid');
  }

  Future<void> _writeConfig(List<int> bytes) async {
    final value = ProtocolCodecs.decodeConfig(bytes);
    if (scenario == FakeBleScenario.busyThenOk && _busyResponses < 3) {
      _busyResponses++;
      _emitResult(DeviceCommandId.configWrite, CommandStatus.busy);
      return;
    }
    if (scenario == FakeBleScenario.rangeError) {
      _emitResult(DeviceCommandId.configWrite, CommandStatus.range, detail: 2);
      return;
    }
    if (scenario == FakeBleScenario.storageError) {
      _emitResult(DeviceCommandId.configWrite, CommandStatus.storage);
      return;
    }
    final validation = ConfigValidator.validate(value);
    if (!validation.isValid) {
      _emitResult(
        DeviceCommandId.configWrite,
        CommandStatus.range,
        detail: validation.issues.first.fieldId,
      );
      return;
    }
    _config = value;
    _emit(BleUuids.configRead, ProtocolCodecs.encodeConfig(_config));
    if (scenario != FakeBleScenario.writeTimeout) {
      _emitResult(DeviceCommandId.configWrite, CommandStatus.ok);
    }
  }

  Future<void> _writeCommand(List<int> bytes) async {
    final command = ProtocolCodecs.decodeCommand(bytes);
    switch (command.id) {
      case DeviceCommandId.resetTrip:
        _telemetry = _telemetry.copyWith(
          speedX100: 0,
          avgSpeedX100: 0,
          maxSpeedX100: 0,
          tripDistanceCm: 0,
          movingTimeS: 0,
          revolutions: 0,
        );
      case DeviceCommandId.displayOn:
        _telemetry = _telemetry.copyWith(flags: _telemetry.flags | 0x02);
      case DeviceCommandId.displayOff:
        _telemetry = _telemetry.copyWith(flags: _telemetry.flags & ~0x02);
      case DeviceCommandId.sensorTestStart:
        _sensorTest = true;
        _restartTelemetryTimer();
      case DeviceCommandId.sensorTestStop:
        _sensorTest = false;
        _restartTelemetryTimer();
      default:
        break;
    }
    _emitResult(command.id, CommandStatus.ok);
    _emit(
      BleUuids.telemetry,
      ProtocolCodecs.encodeTelemetry(_profiledTelemetry()),
    );
  }

  void _emitResult(
    DeviceCommandId command,
    CommandStatus status, {
    int detail = 0,
  }) {
    _emit(
      BleUuids.commandResult,
      ProtocolCodecs.encodeCommandResult(
        CommandResult(
          structVersion: 1,
          commandId: command,
          status: status,
          detail: detail,
          token: 0,
          payload: const <int>[],
        ),
      ),
    );
  }

  void _emit(String uuid, List<int> bytes) {
    final controller = _notifications[uuid];
    if (controller != null && !controller.isClosed) controller.add(bytes);
  }

  @override
  Stream<List<int>> subscribe(String characteristicUuid) => _notifications
      .putIfAbsent(characteristicUuid, StreamController<List<int>>.broadcast)
      .stream;

  Telemetry _profiledTelemetry() {
    final seq = (_telemetry.seq + 1) & 0xFFFF;
    _telemetry = switch (_profile) {
      FakeRideProfile.idle => _telemetry.copyWith(
        flags: 0x02,
        speedX100: 0,
        rideState: RideState.idle,
        sensorState: SensorState.noSignal,
        powerState: PowerState.shortStop,
        lastPulseAgeMs: 0xFFFFFFFF,
        seq: seq,
      ),
      FakeRideProfile.paused => _telemetry.copyWith(
        flags: 0x22,
        speedX100: 0,
        rideState: RideState.paused,
        sensorState: SensorState.idle,
        powerState: PowerState.shortStop,
        lastPulseAgeMs: 5000,
        seq: seq,
      ),
      FakeRideProfile.lowBattery => _telemetry.copyWith(
        flags: 0x33,
        batteryMv: 3420,
        batteryPct: 12,
        seq: seq,
      ),
      FakeRideProfile.moving => _telemetry.copyWith(
        flags: 0x23,
        speedX100: 2550,
        tripDistanceCm: _telemetry.tripDistanceCm + 70,
        movingTimeS: _telemetry.movingTimeS + 1,
        revolutions: _telemetry.revolutions + 1,
        lastPulseAgeMs: 250,
        rideState: RideState.moving,
        sensorState: SensorState.ok,
        powerState: PowerState.active,
        seq: seq,
      ),
    };
    return _telemetry;
  }

  void setRideProfile(FakeRideProfile profile) {
    _profile = profile;
  }

  void _restartTelemetryTimer() {
    _telemetryTimer?.cancel();
    _telemetryTimer = Timer.periodic(
      Duration(milliseconds: _sensorTest ? 200 : 1000),
      (_) {
        if (_connected) {
          _emit(
            BleUuids.telemetry,
            ProtocolCodecs.encodeTelemetry(_profiledTelemetry()),
          );
        }
      },
    );
  }

  @override
  Future<int> readRssi() async {
    _requireConnected();
    return -58;
  }

  @override
  Future<void> dispose() async {
    _telemetryTimer?.cancel();
    await disconnect();
    for (final controller in _notifications.values) {
      await controller.close();
    }
    _notifications.clear();
  }
}

File: web-app/src/transport/fakeBleTransport.ts
import { BleUuids } from '../protocol/uuids';
import {
  decodeCommand,
  decodeConfig,
  encodeCommandResult,
  encodeConfig,
  encodeDeviceInfo,
  encodeTelemetry,
  SIZES,
} from '../protocol/codecs';
import { validateConfig } from '../domain/configValidator';
import {
  type DeviceCommand,
  type DeviceConfig,
  type DeviceInfo,
  type Telemetry,
  defaultConfig,
  type CommandStatus,
  type DeviceCommandId,
} from '../domain/types';
import type { BleDiscovery, BleLinkState, BleScanResult, BleTransport } from './bleTransport';

export type FakeBleScenario =
  | 'normal'
  | 'serviceMissing'
  | 'mtuTooSmall'
  | 'protocolMajor2'
  | 'pairingClosed'
  | 'writeTimeout'
  | 'busyThenOk'
  | 'rangeError'
  | 'storageError'
  | 'disconnectOnWrite';

export type FakeRideProfile = 'idle' | 'moving' | 'paused' | 'lowBattery';

type NotifyHandler = (bytes: Uint8Array) => void;

export class FakeBleTransport implements BleTransport {
  scenario: FakeBleScenario = 'normal';
  private _profile: FakeRideProfile = 'moving';
  private _connected = false;
  private _sensorTest = false;
  private _busyResponses = 0;
  private _pendingDangerousToken = 0;
  private _linkListeners = new Set<(state: BleLinkState) => void>();
  private _notifyHandlers = new Map<string, Set<NotifyHandler>>();
  private _telemetryTimer: ReturnType<typeof setInterval> | null = null;
  readonly operationLog: string[] = [];

  private _config: DeviceConfig = defaultConfig();
  private _telemetry: Telemetry = {
    structVersion: 1,
    flags: 35,
    speedX100: 2550,
    avgSpeedX100: 2200,
    maxSpeedX100: 3500,
    tripDistanceCm: 125000,
    movingTimeS: 1800,
    odometerM: 123456,
    batteryMv: 3900,
    batteryPct: 75,
    rideState: 'moving',
    revolutions: 500,
    lastPulseAgeMs: 250,
    seq: 42,
    sensorState: 'ok',
    powerState: 'active',
    cadenceX10: 0,
    cscFlags: 0,
    lastCrankEventAgeMs: 0xffffffff,
  };

  constructor(profile: FakeRideProfile = 'moving') {
    this._profile = profile;
  }

  get adapterState() {
    return 'ready' as const;
  }

  get connected() {
    return this._connected;
  }

  private get deviceInfo(): DeviceInfo {
    return {
      structVersion: 1,
      protoMajor: this.scenario === 'protocolMajor2' ? 2 : 1,
      protoMinor: 2,
      hwRevision: 1,
      model: 'BIKECOMP-XIAO',
      fwVersion: '1.0.0-fake',
      serial: [1, 2, 3, 4, 5, 6, 7, 8],
      uptimeS: 3600,
      resetReason: 'powerOn',
      bootCount: 42,
      flags: this.scenario === 'pairingClosed' ? 0x0f : 0x1f,
    };
  }

  async *scan(): AsyncIterable<BleScanResult> {
    await delay(40);
    yield {
      deviceId: 'FA:KE:BI:KE:00:01',
      name: 'BikeComp-FAKE',
      rssi: -58,
      bondState: 'bonded',
    };
  }

  async stopScan(): Promise<void> {}

  async *connect(_deviceId: string): AsyncIterable<BleLinkState> {
    const queue: BleLinkState[] = [];
    let resolve: (() => void) | null = null;
    const push = (state: BleLinkState) => {
      queue.push(state);
      resolve?.();
      resolve = null;
    };
    const listener = (state: BleLinkState) => push(state);
    this._linkListeners.add(listener);

    push('connecting');
    await delay(40);
    this._connected = true;
    this._restartTelemetryTimer();
    push('connected');

    try {
      while (this._connected || queue.length > 0) {
        if (queue.length === 0) {
          await new Promise<void>((r) => {
            resolve = r;
          });
        }
        while (queue.length > 0) {
          yield queue.shift()!;
        }
      }
    } finally {
      this._linkListeners.delete(listener);
    }
  }

  async disconnect(): Promise<void> {
    this._connected = false;
    this._stopTelemetryTimer();
    for (const listener of this._linkListeners) {
      listener('disconnected');
    }
  }

  async requestMtu(mtu: number): Promise<number> {
    this.operationLog.push(`mtu:${mtu}`);
    this._requireConnected();
    return this.scenario === 'mtuTooSmall' ? 23 : 247;
  }

  async discoverServices(): Promise<BleDiscovery> {
    this.operationLog.push('discover');
    this._requireConnected();
    if (this.scenario === 'serviceMissing') {
      return { characteristics: new Set() };
    }
    return {
      characteristics: new Set([
        BleUuids.deviceInfo,
        BleUuids.telemetry,
        BleUuids.configRead,
        BleUuids.configWrite,
        BleUuids.command,
        BleUuids.commandResult,
        BleUuids.companionWrite,
        BleUuids.errorLog,
      ]),
    };
  }

  async read(characteristicUuid: string): Promise<Uint8Array> {
    this.operationLog.push(`read:${characteristicUuid}`);
    this._requireConnected();
    switch (characteristicUuid) {
      case BleUuids.deviceInfo:
        return encodeDeviceInfo(this.deviceInfo);
      case BleUuids.telemetry:
        return encodeTelemetry(this._profiledTelemetry());
      case BleUuids.configRead:
        return encodeConfig(this._config);
      case BleUuids.errorLog:
        return new Uint8Array([1, 1, 0, 0, 0, 0, 0, 0, 0, 0]);
      default:
        throw new Error(`Unsupported fake characteristic ${characteristicUuid}`);
    }
  }

  async writeWithResponse(characteristicUuid: string, value: Uint8Array): Promise<void> {
    this._requireConnected();
    if (this.scenario === 'disconnectOnWrite') {
      this._connected = false;
      for (const listener of this._linkListeners) listener('disconnected');
      throw new Error('Injected disconnect');
    }
    if (characteristicUuid === BleUuids.configWrite) {
      await this._writeConfig(value);
      return;
    }
    if (characteristicUuid === BleUuids.companionWrite) {
      this.operationLog.push(`companion:${value.length}`);
      return;
    }
    if (characteristicUuid === BleUuids.command) {
      await this._writeCommand(value);
      return;
    }
    throw new Error(`Unsupported fake write ${characteristicUuid}`);
  }

  async *subscribe(characteristicUuid: string): AsyncIterable<Uint8Array> {
    const handlers = new Set<NotifyHandler>();
    this._notifyHandlers.set(characteristicUuid, handlers);
    const queue: Uint8Array[] = [];
    let wake: (() => void) | null = null;
    const handler: NotifyHandler = (bytes) => {
      queue.push(bytes);
      wake?.();
      wake = null;
    };
    handlers.add(handler);
    try {
      while (true) {
        if (queue.length === 0) {
          await new Promise<void>((r) => {
            wake = r;
          });
        }
        while (queue.length > 0) {
          yield queue.shift()!;
        }
      }
    } finally {
      handlers.delete(handler);
    }
  }

  async readRssi(): Promise<number> {
    return -58;
  }

  private _requireConnected() {
    if (!this._connected) throw new Error('Fake BLE device is disconnected');
  }

  private _emit(uuid: string, bytes: Uint8Array) {
    const handlers = this._notifyHandlers.get(uuid);
    if (!handlers) return;
    for (const h of handlers) h(bytes);
  }

  private _emitResult(id: DeviceCommandId, status: CommandStatus, detail = 0, token = 0) {
    if (this.scenario === 'writeTimeout') return;
    this._emit(
      BleUuids.commandResult,
      encodeCommandResult({
        structVersion: 1,
        commandId: id,
        status,
        detail,
        token,
        payload: [],
      }),
    );
  }

  private async _writeConfig(bytes: Uint8Array) {
    const value = decodeConfig(bytes);
    if (this.scenario === 'busyThenOk' && this._busyResponses < 3) {
      this._busyResponses++;
      this._emitResult('configWrite', 'busy');
      return;
    }
    if (this.scenario === 'rangeError') {
      this._emitResult('configWrite', 'range', 2);
      return;
    }
    if (this.scenario === 'storageError') {
      this._emitResult('configWrite', 'storage');
      return;
    }
    const validation = validateConfig(value);
    if (!validation.isValid) {
      this._emitResult('configWrite', 'range', validation.issues[0].fieldId);
      return;
    }
    this._config = value;
    this._emit(BleUuids.configRead, encodeConfig(this._config));
    this._emitResult('configWrite', 'ok');
  }

  private _handleDangerous(command: DeviceCommand): boolean {
    const dangerous = ['resetOdometer', 'factoryReset', 'reboot', 'setOdometer', 'openPairingWindow'];
    if (!dangerous.includes(command.id)) return false;

    if (!command.hasToken) {
      this._pendingDangerousToken = (Math.random() * 0xffffffff) >>> 0;
      this._emitResult(command.id, 'needsConfirm', 0, this._pendingDangerousToken);
      return true;
    }
    const tokenOk =
      command.payload.length >= 4 &&
      new DataView(Uint8Array.from(command.payload.slice(-4)).buffer).getUint32(0, true) ===
        this._pendingDangerousToken;
    this._emitResult(command.id, tokenOk ? 'ok' : 'tokenInvalid');
    return true;
  }

  private async _writeCommand(bytes: Uint8Array) {
    const command = decodeCommand(bytes);
    if (this._handleDangerous(command)) return;

    switch (command.id) {
      case 'resetTrip':
        this._telemetry = {
          ...this._telemetry,
          speedX100: 0,
          avgSpeedX100: 0,
          maxSpeedX100: 0,
          tripDistanceCm: 0,
          movingTimeS: 0,
          revolutions: 0,
        };
        break;
      case 'displayOn':
        this._telemetry = { ...this._telemetry, flags: this._telemetry.flags | 0x02 };
        break;
      case 'displayOff':
        this._telemetry = { ...this._telemetry, flags: this._telemetry.flags & ~0x02 };
        break;
      case 'sensorTestStart':
        this._sensorTest = true;
        this._restartTelemetryTimer();
        break;
      case 'sensorTestStop':
        this._sensorTest = false;
        this._restartTelemetryTimer();
        break;
      case 'getDiagnostic': {
        const payload = new Uint8Array(SIZES.diagnosticPayload);
        const view = new DataView(payload.buffer);
        view.setUint32(0, 1000, true);
        this._emitResult('getDiagnostic', 'ok', 0, 0);
        this._emit(
          BleUuids.commandResult,
          encodeCommandResult({
            structVersion: 1,
            commandId: 'getDiagnostic',
            status: 'ok',
            detail: 0,
            token: 0,
            payload: Array.from(payload),
          }),
        );
        return;
      }
      default:
        break;
    }
    this._emitResult(command.id, 'ok');
    this._emit(BleUuids.telemetry, encodeTelemetry(this._profiledTelemetry()));
  }

  private _profiledTelemetry(): Telemetry {
    switch (this._profile) {
      case 'idle':
        return { ...this._telemetry, speedX100: 0, rideState: 'idle' };
      case 'paused':
        return { ...this._telemetry, speedX100: 0, rideState: 'paused' };
      case 'lowBattery':
        return {
          ...this._telemetry,
          batteryPct: 12,
          batteryMv: 3400,
          flags: this._telemetry.flags | 0x10,
        };
      default:
        return this._telemetry;
    }
  }

  private _restartTelemetryTimer() {
    this._stopTelemetryTimer();
    const interval = this._sensorTest ? 200 : 1000;
    this._telemetryTimer = setInterval(() => {
      if (!this._connected) return;
      this._telemetry = {
        ...this._telemetry,
        seq: (this._telemetry.seq + 1) % 65536,
      };
      this._emit(BleUuids.telemetry, encodeTelemetry(this._profiledTelemetry()));
    }, interval);
  }

  private _stopTelemetryTimer() {
    if (this._telemetryTimer) {
      clearInterval(this._telemetryTimer);
      this._telemetryTimer = null;
    }
  }
}

function delay(ms: number) {
  return new Promise((r) => setTimeout(r, ms));
}
